course_UvA-DL/activation-functions/Activation_Functions.py

```diff
@@ -0,0 +1,824 @@
+
+
+class LeakyReLU(ActivationFunction):
+
+    def __init__(self, alpha=0.1):
+        super().__init__()
+        self.config["alpha"] = alpha
+
+    def forward(self, x):
+        return torch.where(x > 0, x, self.config["alpha"] * x)
+
+
+##############################
+
+
+class ELU(ActivationFunction):
+
+    def forward(self, x):
+        return torch.where(x > 0, x, torch.exp(x) - 1)
+
+
+##############################
+
+
+class Swish(ActivationFunction):
+
+    def forward(self, x):
+        return x * torch.sigmoid(x)
+
+
+##############################
+
+# %% [markdown]
+# For later usage, we summarize all our activation functions in a dictionary mapping the name to the class object.
+# In case you implement a new activation function by yourself, add it here to include it in future comparisons as well:
+
+# %%
+act_fn_by_name = {
+    "sigmoid": Sigmoid,
+    "tanh": Tanh,
+    "relu": ReLU,
+    "leakyrelu": LeakyReLU,
+    "elu": ELU,
+    "swish": Swish
+}
+
+# %% [markdown]
+# ### Visualizing activation functions
+#
+# To get an idea of what each activation function actually does, we will visualize them in the following.
+# Next to the actual activation value, the gradient of the function is an important aspect as it is crucial for optimizing the neural network.
+# PyTorch allows us to compute the gradients simply by calling the `backward` function:
+
+
+# %%
+def get_grads(act_fn, x):
+    """Computes the gradients of an activation function at specified positions.
+
+    Inputs:
+        act_fn - An object of the class "ActivationFunction" with an implemented forward pass.
+        x - 1D input tensor.
+    Output:
+        A tensor with the same size of x containing the gradients of act_fn at x.
+    """
+    x = x.clone().requires_grad_()  # Mark the input as tensor for which we want to store gradients
+    out = act_fn(x)
+    out.sum().backward()  # Summing results in an equal gradient flow to each element in x
+    return x.grad  # Accessing the gradients of x by "x.grad"
+
+
+# %% [markdown]
+# Now we can visualize all our activation functions including their gradients:
+
+
+# %%
+def vis_act_fn(act_fn, ax, x):
+    # Run activation function
+    y = act_fn(x)
+    y_grads = get_grads(act_fn, x)
+    # Push x, y and gradients back to cpu for plotting
+    x, y, y_grads = x.cpu().numpy(), y.cpu().numpy(), y_grads.cpu().numpy()
+    # Plotting
+    ax.plot(x, y, linewidth=2, label="ActFn")
+    ax.plot(x, y_grads, linewidth=2, label="Gradient")
+    ax.set_title(act_fn.name)
+    ax.legend()
+    ax.set_ylim(-1.5, x.max())
+
+
+# Add activation functions if wanted
+act_fns = [act_fn() for act_fn in act_fn_by_name.values()]
+x = torch.linspace(-5, 5, 1000)  # Range on which we want to visualize the activation functions
+# Plotting
+rows = math.ceil(len(act_fns) / 2.0)
+fig, ax = plt.subplots(rows, 2, figsize=(8, rows * 4))
+for i, act_fn in enumerate(act_fns):
+    vis_act_fn(act_fn, ax[divmod(i, 2)], x)
```

Comment:
```suggestion
cols = 2
rows = math.ceil(len(act_fns) / float(cols))
fig, ax = plt.subplots(rows, cols, figsize=(cols * 4, rows * 4))
for i, act_fn in enumerate(act_fns):
    vis_act_fn(act_fn, ax[divmod(i, cols)], x)
```
to be more generic and reader can easier trace the meaning of 2